state/domain_mem.go

```diff
@@ -0,0 +1,443 @@
+package state
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/ledgerwatch/erigon-lib/commitment"
+	"github.com/ledgerwatch/erigon-lib/common"
+	"github.com/ledgerwatch/erigon-lib/common/length"
+	"github.com/ledgerwatch/erigon-lib/etl"
+)
+
+type DomainMem struct {
+	*Domain
+
+	etl    *etl.Collector
+	mu     sync.RWMutex
+	values map[string]*KVList
+	latest map[string][]byte // key+^step -> value
+}
+
+type KVList struct {
+	TxNum []uint64
+	//Keys  []string
+	Vals [][]byte
+}
+
+func (l *KVList) Latest() (tx uint64, v []byte) {
+	sz := len(l.TxNum)
+	if sz == 0 {
+		return 0, nil
+	}
+	sz--
+
+	tx = l.TxNum[sz]
+	v = l.Vals[sz]
+	return tx, v
+}
+
+func (l *KVList) Put(tx uint64, v []byte) (prevTx uint64, prevV []byte) {
+	prevTx, prevV = l.Latest()
+	l.TxNum = append(l.TxNum, tx)
+	l.Vals = append(l.Vals, v)
+	return
+}
+
+func (l *KVList) Len() int {
+	return len(l.TxNum)
+}
+
+func (l *KVList) Apply(f func(txn uint64, v []byte) error) error {
+	for i, tx := range l.TxNum {
+		if err := f(tx, l.Vals[i]); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (l *KVList) Reset() {
+	//l.Keys = l.Keys[:0]
```

Comment:
I prefer as the first step: get much green tests. And ten introduce prealloc/reset optimizations.